### DISCUSSION TITLE
Umi & Bigfish 想法征集 08

### DISCUSSION THREAD

**Discussion by sorrycc** [umijs/umi#10946] [Ideas]
> Why this issue? 计划在 Umi Contributor 群增加一条规则。每人每周贡献 2 个改进 Umi 或 Bigfish 的想法（辟起、辟殊、云谦），外包同学减半（1 个），外部 Contributor 随意，我每周三会在群里发个 discussion 的链接做收集。

直接回复即可。（含内部敏感信息的可以私发我）

**Comment by sorrycc** [comment]
- 基于 magicast 改造 umi config（但要注意 magicast 好像不支持找 scope bindings）
- 基于 magicast 改造 codemod 实现
- Bun 0.6 即将发布，学习下他做 Bundle 的思路
- did-you-know 找 cache 路径那有点问题，可能会不存在
- .umi 下的类型可以把配置项里 define 里定义的全局变量带上
- （内部）写个脚本根据用户描述的问题自动初始化环境并启动（包括 clone 代码、安装依赖、启动 dev 等）
- （内部）Bigfish 官网搞个服务，`POST /api/reports`，接受 JSON 格式的数据，然后传导 oss 上，返回具体地址。这个服务可用于反馈问题时上报数据，可 @alipay/bigfish-reporter 结合。内容包含脚本信息、日志信息、Bigfish Doctor 信息。（现在填了已跑 Doctor，但实际上跑的怎么样，我们并不清楚）
- （内部）Bigfish 3 升 4 的 CodeMod，自动处理 `pages/**/_mock.(t|j)s` 的 mock 文件，移动到 `src/mock` 目录下。

**Reply by PeachScript** [reply]
这一看就是屯了两周的 👀

**Comment by PeachScript** [comment]
- `enhanced-resolve` 在 Umi 和 dumi 里都有多处使用，可以放在 bundler-utils 里？
- Umi 发布脚本的后置 sync 任务前可以加一个轮询检测逻辑，等 npm tag 更新后再执行 sync，避免只 sync 到部分包或无效 sync
- Umi 发布脚本可以加一个前置任务，检查 status.npmjs.com 各项能力是否处于正常状态，降低发布风险（比如 npm tag 死活不更新）
- `api.config` 可以用 `defineProperty` 来控制调用时机，避免插件开发者在钩子外或部分钩子内提前访问，导致功能异常

**Reply by fz6m** [reply]
第二条可以检测到发布 tag 后在 github action 里做，不用等本地轮训了。

**Comment by fengsx** [comment]
rspack 或者 turpopack？

**Comment by xierenyuan** [comment]
Umi 切到 es module 支持 (原因社区愈来愈多包 只支持 es module 了 在 umi 插件中使用 现在有两种路径 1 是 预打包  2 是 通过 import() 有一些成本) 长期来看 切到 es module 是趋势

**Reply by fz6m** [reply]
nextjs 目前不会切 esm native 的话，我认为没必要切，未来如果有重量级选手表态后可以跟。

**Comment by xiaohuoni** [comment]
plugin-access 和 model，initial State，layout 插件强制捆绑。不用其中的任一个插件，都无法完整使用 access 功能，感觉和设计不符合 - 所有的插件都是可拔插的。

## initial State

plugin-access 是不是应该更干净，比如

```
// src/access.ts
export default function (initialState) {
  const { userId, role } = initialState;
 
  return {
    canReadFoo: true,
    canUpdateFoo: role === 'admin',
    canDeleteFoo: (foo) => {
      return foo.ownerId === userId;
    },
  };
}
```

应该设计成 onInitialState ，然后用户可选数据流，在 onInitialState return initialState

## layout

另外 layout 配置也很怪。

```
export const layout: RunTimeLayoutConfig = () => {
  return {
    // 自定义 403 页面
    unAccessible: <div>'unAccessible'</div>,
    // 自定义 404 页面
    noFound: <div>'noFound'</div>,
  };
};
```

可能是独立的 unAccessible 和 noFound 函数，然后在 layout 插件中去响应这两个函数，return 回需要的内容。

**Comment by hanzebang** [comment]
基于上次的投诉事件和上月末的二线答疑情况，在想要不要建立一个「二线答疑日」，比如每两周选出1天由二线用户单人轮流或者三人共同进行答疑，大致考虑了以下几种情况
- 可以方便处理一些需求、个别用户想直接对话二线深入了解框架等情况
- 便于发现一线同学/流程上有哪些不足或者可以改进的地方
- 便于发现用户对框架代码、提示、文档等感到困惑的问题，以及当下热点问题等

**Reply by sorrycc** [reply]
感觉可以，参与下一线答疑还是必要的。

**Comment by xiaohuoni** [comment]
现在关于 umi 的用户，只有说明文档里面的"同时也服务了不少外部用户，包括淘系、飞猪、阿里云、字节、腾讯、口碑、美团等。"
是不是可以搞一个 showcase 类似 https://nextjs.org/showcase

**Reply by sorrycc** [reply]
目前没有做收集，可以先找个地方收集下。

**Reply by xiaohuoni** [reply]
https://github.com/umijs/umi/issues/10964 Umi Users 👨🏻‍💻👩🏻‍💻👨🏻‍💻👩🏻‍💻

**Comment by xiaohuoni** [comment]
如果用户需要写全局样式(:root { --adm-color-primary: red })，应该要写在哪？写在 global.less  dev 有效 build 之后顺序也不对。config 的 styles 也是优先级低于项目内用户编写样式。

**Comment by kurisu994** [comment]
希望能支持rspack

**Comment by xXAvoraXx** [comment]
由于 umi 的静态路由定义无法解决我们的问题，因此我们需要 api 的动态路由定义。我已经开始为这种情况开发一个插件，我们可以一起开发。

https://github.com/umijs/umi/discussions/11770

**Reply by xXAvoraXx** [reply]
计划中应包括将 umijs 文件翻译成英文，使其能够与其他竞争解决方案竞争，并在行业中占有一席之地。
